### PR TITLE
Elasticsearch v2 integration

### PR DESCRIPTION
**Pull request by roysagi**
## Status
Ready

## Related Issues
Fixes: https://github.com/demisto/etc/issues/18446
Fixes: https://github.com/demisto/etc/issues/15454
Fixes: https://github.com/demisto/etc/issues/11999

## Description
Adding new integration: Elasticsearch v2 - now supporting versions 6 and 7.

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

## Additional changes
New docker image was pushed to fit this integration

### CODE REVIEW

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+from elasticsearch import Elasticsearch, RequestsHttpConnection
+from elasticsearch_dsl import Search
+from elasticsearch_dsl.query import QueryString
+from datetime import datetime
+import json
+import requests
+
+
+SERVER = demisto.params()['url'][:-1] if demisto.params()['url'].endswith('/') else demisto.params()['url']
+USERNAME = demisto.params()['username']
```
Comment (Itay4) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: use credentials

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        for source_field in hit.get('_source').keys():
+            if fields is None or source_field in fields:  # filtering only the requested fields to display
+                table_context[str(source_field)] = hit.get('_source').get(str(source_field))
+                headers.append(source_field)
+
+    return table_context, headers
+
+
+def search_command():
+    index = demisto.args()['index']
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: only use .get for demisto.args

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    query = demisto.args()['query']
+    fields = demisto.args().get('fields')  # fields to display
+    explain = 'true' == demisto.args().get('explain')
+    base_page = int(demisto.args().get('page'))
+    size = int(demisto.args().get('size'))
+    sort_field = demisto.args().get('sort-field')
+    sort_order = demisto.args().get('sort-order')
+
+    if USERNAME is None:
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: Same here

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        es = Elasticsearch(hosts=[SERVER], connection_class=RequestsHttpConnection, verify_certs=INSECURE)
+
+    else:
+        es = Elasticsearch(hosts=[SERVER], connection_class=RequestsHttpConnection,
+                           http_auth=(USERNAME, PASSWORD), verify_certs=INSECURE)
+
+    q = QueryString(query=query)
+    s = Search(using=es, index=index).query(q)[base_page:base_page + size]
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]:
```suggestion
    s = Search(using=es, index=index).query(q)[base_page:base_page + size + 1]
```

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    if explain:
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: add explanations

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    if hit.get('_source') is not None:
+        for source_field in hit.get('_source').keys():
+            if fields is None or source_field in fields:  # filtering only the requested fields to display
+                table_context[str(source_field)] = hit.get('_source').get(str(source_field))
+                headers.append(source_field)
+
+    return table_context, headers
+
+
+def search_command():
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: separate into several functions, it's way too long

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        for hit in response.get('hits').get('hits'):
+            single_hit_table, single_header = get_hit_table(hit, fields)
+            hit_tables.append(single_hit_table)
+            hit_headers = list(set(single_header + hit_headers) - set(['_id', '_type', '_index', '_score']))
+        hit_headers = ['_id', '_index', '_type', '_score'] + hit_headers
+
+    search_context['Results'] = response.get('hits').get('hits')
+    meta_headers = ['Query', 'took', 'timed_out', 'total', 'max_score', 'Server', 'Page', 'Size']
+
+    return {
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: use return_outputs

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+                        'sourceBrand': 'Elasticsearch',
+                        'name': 'Elasticsearch: Index: ' + str(FETCH_INDEX) + ", ID: " + str(hit.get('_id')),
+                        'details': json.dumps(hit.get('_source')),
+                        'rawJSON': json.dumps(hit),
+                        'labels': incident_label_maker(hit.get('_source')),
+                        'occurred': temp_date.isoformat() + 'Z'
+                    }
+                    incidents.append(inc)
+
+        demisto.info('extract {} incidents'.format(len(incidents)))
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: I do get incidents and it is the same as in Gmail - I can demo it for you

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 4 above)
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: The use is without the +1

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    handle_proxy()
+    if demisto.command() == 'test-module':
+        test_func()
+    elif demisto.command() == 'fetch-incidents':
+        fetch_incidents()
+    elif demisto.command() == 'search':
+        search_command()
+except Exception as e:
+    return_error(str(e))
+    LOG.print_log()
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: remove this line

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ronykoz) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: @yaakovi what am i missing here

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    try:
+        res = requests.get(SERVER, auth=(USERNAME, PASSWORD), verify=INSECURE)
+        if res.status_code >= 400:
+            return_error("Failed to connect, Got the following error: " + HTTP_ERRORS[int(res.status_code)])
+
+    except requests.exceptions.RequestException:
+        return_error("Failed to connect, Check Server URL and Port number")
+
+    if demisto.params().get('isFetch'):
+        str_error = ''
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: change to 
```suggestion
        str_error = []
```

then do

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        if TIME_FIELD is None:
+            str_error = str_error + "\nNo time field for fetch"
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]:
```suggestion
            str_error.append("No time field for fetch")
```

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+            str_error = str_error + "\nNo index for fetch"
+
+        if FETCH_QUERY is None:
+            str_error = str_error + "\nNo query for fetch"
+
+        if TIME_FORMAT is None:
+            str_error = str_error + "\nNo time format for fetch"
+
+        if str_error != '':
+            return_error("Got The following errors in test: " + str_error)
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]:
```suggestion
            return_error("Got The following errors in test:\n" + '\n'.join(str_error))
```

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+            query = QueryString(query=str(TIME_FIELD) + ":*")
+            search = Search(using=es, index=FETCH_INDEX).query(query)[0:1]
+            response = search.execute().to_dict()
+            _, total_results = get_total_results(response)
+            if total_results > 0:
+                hit_date = str(response.get('hits', {}).get('hits')[0].get('_source').get(str(TIME_FIELD)))
+                datetime.strptime(hit_date, TIME_FORMAT).isoformat() + 'Z'
+
+        except ValueError as e:
+            return_error("Inserted time format does not match. " + str(e))
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: Maybe like this?
```suggestion
            return_error("Inserted time format does not match your index. " + str(e))
```

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+    except requests.exceptions.RequestException:
+        return_error("Failed to connect, Check Server URL and Port number")
+
+    if demisto.params().get('isFetch'):
+        str_error = ''
+        if TIME_FIELD is None:
+            str_error = str_error + "\nNo time field for fetch"
+
+        if FETCH_INDEX is None:
+            str_error = str_error + "\nNo index for fetch"
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]:
```suggestion
            str_error = str_error + "Index is not configured for fetch in the integration parameters"
```

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+        search = search.source(fields)
+
+    if sort_field is not None:
+        search = search.sort({sort_field: {'order': sort_order}})
+
+    response = search.execute().to_dict()
+    total_dict, total_results = get_total_results(response)
+    search_context, meta_headers, hit_tables, hit_headers = results_to_context(index, query, base_page,
+                                                                               size, total_dict, fields, response)
+    return_outputs(tableToMarkdown('Search Metadata:', search_context, meta_headers, removeNull=True) + '\n'
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: Separate into several lines, assign arguments - make it more readable

**Comment by ronykoz** [comment]
@kirbles19 could you please go over:
https://github.com/demisto/content/blob/Elasticsearch_integration/Integrations/Elasticsearch_v2/Elasticsearch_v2.yml

https://github.com/demisto/content/blob/Elasticsearch_integration/Integrations/Elasticsearch_v2/CHANGELOG.md

And please go over the error msgs with @roysagi

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yaakovi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: `demisto.log`, `print`, `demisto.results` will fail and cause issues while running under `fetch-incident` and `long-running` scope.

`demisto.error`, demisto.info`, `LOG` (and recently `return_error` as well) works in every integration scope.

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.yml [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+      description: The page from which the search started.
+      type: Number
+    - contextPath: Elasticsearch.Search.Size
+      description: The maximal amount of scores that can return.
+      type: Number
+  dockerimage: demisto/elasticsearch:1.0.0.1795
+  isfetch: true
+  runonce: false
+  script: '-'
+  type: python
```
[the remark this answers is not in the document]
Comment (ronykoz) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.yml]: keep this

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+    # handle first time fetch
+    if last_fetch is None:
+        last_fetch = datetime.now() - timedelta(days=FETCH_DAYS)
+
+    else:
+        last_fetch = datetime.strptime(last_fetch, '%Y-%m-%dT%H:%M:%SZ')
+
+    current_fetch = last_fetch
+
+    if PROXY:
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: You use this bundle of rows 3 times in the code, lets convert it into a function

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+            hit_date = datetime.strptime(hit_date, '%Y-%m-%dT%H:%M:%SZ')
+            # update last run
+            if hit_date > last_fetch:
+                last_fetch = hit_date
+
+            # avoid duplication due to weak time query
+            if hit_date > current_fetch:
+                inc = {
+                    'type': 'Elasticsearch',
+                    'sourceBrand': 'Elasticsearch',
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]:
```suggestion
                    'sourceBrand': 'Elasticsearch v2',
```

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+                    'name': 'Elasticsearch: Index: ' + str(FETCH_INDEX) + ", ID: " + str(hit.get('_id')),
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]:
```suggestion
                    'name': 'Elasticsearch v2: Index: ' + str(FETCH_INDEX) + ", ID: " + str(hit.get('_id')),
```

**Comment by ronykoz** [comment]
@ygerowitz please note that @roysagi has continued to work on the yml so please sync before committing anything to the repo/or starting your work on this

**Comment by ygerowitz** [comment]
> @ygerowitz please note that @roysagi has continued to work on the yml so please sync before committing anything to the repo/or starting your work on this

@ronykoz @roysagi Roy told me that he made some changes yesterday, and I'm comparing the files and integrating the changes. I will be in touch with him as well.

**Comment by ronykoz** [comment]
@ygerowitz please don't close PRs.

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+        str_error.append("Query by which to fetch incidents is not configured.")
+
+    if TIME_FORMAT == '' or TIME_FORMAT is None:
+        str_error.append("Time format is not configured.")
+
+    if len(str_error) > 0:
+        return_error("Got the following errors in test:\nFetches incidents is enabled.\n" + '\n'.join(str_error))
+
+
+def test_general_query(es, que, test_num):
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: test_num bad name, please update it

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 21]
```diff
@@ -0,0 +1,315 @@
...
+    if demisto.params().get('isFetch'):
+        # check the existence of all necessary fields for fetch
+        fetch_params_check()
+
+        try:
+            # build general Elasticsearch class
+            es = elasticsearch_builder()
+
+            # test if FETCH_INDEX exists
+            test_general_query(es, '*', 1)
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]: seems strange that you are using the same function for 3 different checks, please update this to separate functions

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.yml [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+      required: false
+      secret: false
+    - default: false
+      defaultValue: '100'
+      description: The number of documents displayed per page. Can be "1" to "10,000". The default is "100".
+      isArray: false
+      name: size
+      required: false
+      secret: false
+    - default: false
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.yml]: no, we currently support a single sort field

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.yml [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+- defaultvalue: 'false'
+  display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+- display: Use system proxy settings
+  name: proxy
+  required: false
+  type: 8
+- display: Index from which to fetch incidents (CSV)
```
[the remark this answers is not in the document]
Comment (roysagi) [reply Integrations/Elasticsearch_v2/Elasticsearch_v2.yml]: yeah

File: Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 24]
```diff
@@ -0,0 +1,348 @@
...
+    demisto.incidents(incidents)
+
+
+try:
+    LOG('command is %s' % (demisto.command(),))
+    if demisto.command() == 'test-module':
+        test_func()
+    elif demisto.command() == 'fetch-incidents':
+        fetch_incidents()
+    elif demisto.command() in ['search', 'elasticsearch-search']:
```
Comment (ronykoz) [inline Integrations/Elasticsearch_v2/Elasticsearch_v2.py]:
```suggestion
    elif demisto.command() in ['search', 'es-search']:
```

**Pull request merged**